src/boh_app/data/generate_recipes.py

```diff
@@ -85,3 +84,29 @@ def mk_model_data(self, recipe: dict[str, Any]) -> None:
             model["skills"] = [self._get_skill(recipe)]
             model["recipe_internals"] = [internal_id]
             self.recipe_map[recipe_key] = model
+
+
+class BookHandler:
+    def __init__(self):
+        self.items = get_valid_refs("item")
+
+    def mk_model_data(self, book: dict[str, Any]) -> Recipe:
+        principle, amount, product = None, None, None
+        for aspect, value in book["aspects"].items():
+            aspect = aspect.removeprefix("mystery.")
+            if aspect in dir(Principle):
+                principle = Principle(aspect)
+                amount = value
+        for name, products in book["xtriggers"].items():
+            if name.startswith("reading."):
+                for p in products:
+                    if p["id"] in self.items:
+                        product = ItemRef(id=p["id"])
```

Comment:
Those are search loops right?

I’d either do

```py
try:
    thing = next(x for xs in y if xs.blah for x in xs)
except StopIteration:
    raise RuntimeError(f"No x found in y: {some_info}")
```

(if not finding things is a valid outcome, just do `thing = next((...), None)`

or, if comprehensions are not powerful enough:

```py
for ...:
    if element.valid:
        thing = element.xy
        break
else:
    raise RuntimeError(...)
```

The `else` branch is hit if one doesn’t `break`